### ISSUE TITLE
Provisions nodes as ipv4 instead of ipv6 when there are control plane issues

### ISSUE THREAD

**Issue by pelzerim** [kubernetes-sigs/karpenter#1384]
### Description

**Observed Behavior**:

Karpenter created ipv4 nodes in a ipv6 EKS cluster.

The nodeclaims were unable to get into status `Initialized`:
```
  Conditions:
    Last Transition Time:  2024-07-01T23:14:18Z
    Message:               KnownEphemeralTaint "node.cloudprovider.kubernetes.io/uninitialized=true:NoSchedule" still exists
    Reason:                KnownEphemeralTaintsExist
    Status:                False
    Type:                  Initialized
    Last Transition Time:  2024-07-01T23:12:20Z
```

The cloud controller was unable to provide node information:

```
I0702 07:23:04.241540      12 node_controller.go:229] error syncing 'ip-10-0-100-244.ap-northeast-2.compute.internal': failed to get node modifiers from cloud provider: provided node ip for node "ip-10-0-100-244.ap-northeast-2.compute.internal" is not valid: failed to get node address from cloud provider that matches ip: 10.0.100.244, requeuing
```

Inspecting the user data of the created ec2 instances reveals that the instances are missing these flags (which are there for older nodes):

```
--ip-family ipv6 \
--dns-cluster-ip 'fdac:***:****::a' \
```

Killing karpenter pods and removing all stuck nodeclaims by hand did resolve the issue.

This AWS EKS cluster is known to have performance issues with the control plane. We have already seen [this](https://github.com/aws/karpenter-provider-aws/issues/4168) issue.  We are dealing with AWS support to get it fixed. 

A bit of digging reveals that the decision to add the flags is [dynamic](https://github.com/aws/karpenter-provider-aws/blob/e8a345723c8db785bd07b8595c395edbdfb9255b/pkg/providers/amifamily/bootstrap/eksbootstrap.go#L118). 

**Expected Behavior**:

Karpenter should be able to provision IPV6 nodes even if the control plane is temporarily unavailable.

Having control over this setting directly would also be very useful?

**Versions**:
- Karpenter CRD Chart Version: 0.37.0
- Karpenter Chart version: 0.37.0
- Kubernetes Version (`kubectl version`): v1.30.0-eks-036c24b
- AWS terraform-aws-modules/eks/aws//modules/karpenter version: 20.8.5

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by jigisha620** [comment]
Have you configured [metadataOptions](https://karpenter.sh/v0.37/concepts/nodeclasses/#specmetadataoptions) on your EC2nodeClass?  This gives you an option to enable IPv6 endpoint for your instances. This option is disabled by default. Can you share what your EC2nodeClass looks like?

**Comment by pelzerim** [comment]
@jigisha620 Thanks for the prompt response! We do not have that option enabled:

```
apiVersion: karpenter.k8s.aws/v1beta1
kind: EC2NodeClass
metadata:
  name: ${name}
spec:
  amiFamily: AL2
  role: ${role}
  subnetSelectorTerms:
    %{ for subnet_id in subnet_ids }
    - id: "${subnet_id}"
    %{ endfor }
  securityGroupSelectorTerms:
    - tags:
        karpenter.sh/discovery: ${cluster_name}
  tags:
    karpenter.sh/discovery: ${cluster_name}
  amiSelectorTerms:
    - id: ${ami_id}
  blockDeviceMappings:
    - deviceName: /dev/xvda
      ebs:
        volumeType: gp3
        volumeSize: ${disk_size_gi}Gi
        deleteOnTermination: true
        encrypted: true
```

Will the change to the metadata service affect the instance kubelet configuration? The corresponding code seems to use ClusterDNS IP: https://github.com/aws/karpenter-provider-aws/blob/e8a345723c8db785bd07b8595c395edbdfb9255b/pkg/providers/amifamily/bootstrap/eksbootstrap.go#L122

**Comment by jigisha620** [comment]
Hi @pelzerim,
You are right. I misunderstood. Wondering if you have specified ClusterDNS via `spec.kubeletConfiguration` since we rely on clusterDNS to pass `--ip-family ipv6`. If not, were you passing it earlier as I can see it from the information you have shared earlier.

**Comment by pelzerim** [comment]
@jigisha620 Sadly we could not find relevant logs (errors) form karpenter from around the time of the incident.

We however had kubernetes api throttling issues in the past with that cluster and i strongly suspect something in that direction. I've stumbled upon this `FlowScheme` [here](https://karpenter.sh/docs/getting-started/getting-started-with-karpenter/#preventing-apiserver-request-throttling) and we indeed do **not** run karpenter in the `kube-system` namespace. I've added the FlowScheme manually now and lets see if this goes away.
